### PR TITLE
Add synonyms for special classes of types.

### PR DESCRIPTION
**Pull request by kosmikus**
See #44.

### CODE REVIEW

**Comment by kosmikus** [comment]
New type synonyms should be re-exported from the main module.

**Comment by phadej** [comment]
:+1:

File: src/Generics/SOP/Universe.hs [hunk 1]
```diff
@@ -152,3 +153,34 @@ class HasDatatypeInfo a where
...
+-- constructor.
+--
+type IsProductType (a :: *) (xs :: [*]) =
+  (Generic a, Code a ~ '[ xs ])
+
+-- | Constraint that captures that a datatype is an enumeration type,
+-- i.e., none of the constructors have any arguments.
+--
+type IsEnumType (a :: *) =
+  (Generic a, All ((~) '[]) (Code a))
```
Comment (phadej) [inline src/Generics/SOP/Universe.hs]: Hmm, should this be also `IsEnumType (a :: *) (xss :: [[*]]) =` ?

Comment (kosmikus) [inline src/Generics/SOP/Universe.hs]: Personally, I think not, as `xss` is already available as `Code a`. But if you disagree, you can try to convince me.

Comment (kosmikus) [inline src/Generics/SOP/Universe.hs]: An additional argument for "no" is that the list in this case is already restricted to be a list of empty lists, so in particular, there are no inner element types anywhere. It's therefore unlikely there will be additional `All` or `All2` constraints. Introducing the `xss` variable just for it to be unused most of the time seems more confusing than helpful.

Comment (phadej) [inline src/Generics/SOP/Universe.hs]: I think that the "consistent" variant would be
```haskell
type IsEnumType (a :: *) (n :: Nat) = (Generic a, All ((~) '[]) (Code a), Length (Code a) ~ n))
```
as `n :: Nat` explains the "shape" of enum type; but that's doesn't fit the `generics-sop`, so the current variant is good.

Comment (kosmikus) [inline src/Generics/SOP/Universe.hs]: Yes, I agree that the `Nat` variant would be consistent. If there's ever a desire to have this, I wouldn't be opposed to change it.

**Pull request merged**